### PR TITLE
add page option when listing gists

### PR DESCRIPTION
**Pull request by kapadia**
Addressing pagination of gists when listing https://github.com/defunkt/gist/issues/201. Apologies if my use of optparse isn't the best.

/cc @ConradIrwin

### CODE REVIEW

**Comment by ConradIrwin** [comment]
@kapadia Thanks for this!

I'd prefer this not to be an option, instead we should just automatically paginate and show all results in the list.

That should be pretty easy to implement, but it'll be fun to test it in the case where you do: `gist -l | head -n 10`, as it'd be nice to exit gracefully in that case (instead of throwing an exception).

**Comment by kapadia** [comment]
@ConradIrwin I changed it up a bit so that pagination automatically happens. It would be nice to send out concurrent requests, but I don't know the Ruby way to do that. You'll find a new method, `list_all_gists`, which you may choose to replace to `list_gists` if you like.

Just to note, I've been wanting a facility to quickly skim through all my gists (public and private), without the need to hit up gist.github.com in the browser. I have many large GeoJSON gists, which severely slows down the page renders while navigating the site. This PR is a nice convenience.

File: lib/gist.rb [hunk 1]
```diff
@@ -181,6 +184,55 @@ def list_gists(user = "")
...
+    end
+
+  end
+
+  def get_gist_pages(url)
+
+    request = Net::HTTP::Get.new(url)
+    response = http(api_url, request)
+
+    m = response.header['link'].match(/page=(\d+)>; rel="last/)
```
Comment (ConradIrwin) [inline lib/gist.rb]: @kapadia I think you should use the rel="next" link header rather than constructing the URLs yourself: https://developer.github.com/v3/#pagination

**Comment by ConradIrwin** [comment]
Looks great otherwise, thanks!

File: lib/gist.rb [hunk 2]
```diff
@@ -154,14 +157,14 @@ def multi_gist(files, options={})
...
-  def list_gists(user = "")
+  def list_gists(user="", page=1)
     url = "#{base_path}"
 
     if user == ""
       access_token = auth_token()
       if access_token.to_s != ''
-        url << "/gists?access_token=" << CGI.escape(access_token)
 
+        url << "/gists?access_token=" << CGI.escape(access_token) << "&page=" << page.to_s
```
Comment (ConradIrwin) [inline lib/gist.rb]: @kapadia Can revert this change now?

File: lib/gist.rb [hunk 3]
```diff
@@ -2,6 +2,9 @@
 require 'cgi'
 require 'uri'
 
+require 'pry'
+
+
```
Comment (ConradIrwin) [inline lib/gist.rb]: :) Also needs to go

File: lib/gist.rb (hunk 2 above)
Comment (kapadia) [inline lib/gist.rb]: yep, reverted.

File: lib/gist.rb (hunk 3 above)
Comment (kapadia) [inline lib/gist.rb]: removed. btw, thanks for pry :)

**Comment by kapadia** [comment]
@ConradIrwin How's this looking? One reason I was keen to construct the urls was in hopes of sending concurrent requests to all pages (though I don't know the Ruby way to do this). Anyway, we're now using the link header to cycle through all pages, and it works.

**Comment by ConradIrwin** [comment]
This looks great! Thanks for being responsive to feedback :).

I've merged and released it as `gist v4.4.0`.

**Comment by kapadia** [comment]
Np. Thanks for the merge.
